### PR TITLE
4.x : Workspace refactoring: feature to add a model in the workspace graph

### PR DESCRIPTION
**Pull request by deepakrkris**
Add a model to a workspace graph
- [X]  - add classes for Graph and Node
- [X]  - add a class for Model which extends Node
- [X]  - add a class for Workspace which extends Graph
- [X]  - add a workspace manager
- [X]  - add a 'ModelDefinition' loopback model 
- [X]  - configure the 'memory' connector with workspace operations
- [X]  - add a Task class to mixin atomic tasks in the workspace
- [X]  - add a gherkin cucumber test to add a model in the workspace graph
- [X] - update server.js to initialize the connector

### CODE REVIEW

**Comment by deepakrkris** [comment]
@slnode test please

File: component/datamodel/graph.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+export class Node {
+  constructor(graph, domain, name, data) {
+    this._graph = graph;
+    this._name = name;
+    this._content = data;
+    this._domain = domain;
+  }
+}
```
Comment (superkhau) [inline component/datamodel/graph.js]: Why are these in the same file? Can we put Node in node.js and import to graph.js for usage?

File: common/models/model-definition.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// Copyright IBM Corp. 2015,2016. All Rights Reserved.
+// Node module: loopback-workspace
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+'use strict';
+var clone = require('lodash').clone;
+
+module.exports = function(ModelDefinition) {
+  /**
+   * creates a model definition.
```
Comment (superkhau) [inline common/models/model-definition.js]: c -> C

File: connector/connector.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+var connector = app.dataSources.db.connector;
+var WorkspaceManager = require('../component/workspace-manager.js');
+
+/**
+ * @class Connector
+ *
+ * performs CRUD operations on the Workspace graph.
+ */
+connector.createModel = function(id, data, cb) {
+  var workspace = WorkspaceManager.getWorkspace();
```
Comment (ritch) [inline connector/connector.js]: How does this work...? I'd expect it to take a workspace id or instance of something...

File: common/models/model-definition.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+   */
+  ModelDefinition.on('dataSourceAttached', function(eventData) {
+    var connector = ModelDefinition.getConnector();
+    ModelDefinition.create = function(data, options, cb) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = null;
+      }
+      var id = data.id;
+      connector.createModel(id, data, function(err, modelDef) {
```
Comment (superkhau) [inline common/models/model-definition.js]: `connector.createModel(id, data, cb);`

File: connector/connector.js (hunk 3 above)
Comment (deepakrkris) [inline connector/connector.js]: for now only one workspace, workspace manager will be modified in a following PR to manage multiple workspaces

Comment (ritch) [inline connector/connector.js]: Looks like something is hardcoded or making an assumption on how to get the workspace instance

File: common/models/model-definition.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      "json": false
+    },
+    "name": {
+      "type": "string",
+      "required": true
+    },
+    "description": {
+      "type": "string"
+    },
+    "plural": "string",
```
Comment (superkhau) [inline common/models/model-definition.json]: strings?

File: component/datamodel/graph.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+/**
+ * @class Graph
+ *
+ * graph class to hold nodes and links.
```
Comment (superkhau) [inline component/datamodel/graph.js]: g -> G

File: component/datamodel/graph.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ */
+module.exports.Graph = class Graph {
```
Comment (superkhau) [inline component/datamodel/graph.js]:
```
module.exports = Graph;

class Graph {
  ...
}
```

File: connector/connector.js (hunk 3 above)
Comment (deepakrkris) [inline connector/connector.js]: when we modify the rest API to have workspace Id included in the URI, then we have to comeback here and provide the id

File: component/datamodel/graph.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    return this._cache[domain][name];
+  }
+};
+
+/**
+ * @class Node
+ *
+ * node to represent an entity.
+ */
+module.exports.Node = class Node {
```
Comment (superkhau) [inline component/datamodel/graph.js]: Move to node.js file and export like above graph comment.

File: component/datamodel/model.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+var Graph = require('./graph.js').Graph;
+var Node = require('./graph.js').Node;
+
+/**
+ * @class Model
+ *
+ * represents a Model artifact in the Workspace graph.
+ */
+module.exports.Model = class Model extends Node {
```
Comment (superkhau) [inline component/datamodel/model.js]:
```
module.exports = Model;

class Model extends Node {

File: component/tasks.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+var Model = require('./datamodel/model.js').Model;
+
+/**
+ * @class Tasks
+ *
+ * Atomic tasks that link the in-memory graph with create/update/delete workspace operations.
+ * Every task can be performed using a processor.
+ */
+module.exports = class Tasks {
```
Comment (superkhau) [inline component/tasks.js]:
```
module.exports = Task;

class Task {
}

File: test/acceptance-tests/workspace-devtime/step-definitions/createModel.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    testsuite.modelName = modelName;
+    ModelDefinition.create(model, {}, function(err, data) {
+      if (err) return next(err);
+      testsuite.modelDef = model;
+      next();
+    });
+  });
+
+  this.Then(/^the model definition is created$/, function(next) {
+    var expect = chai.expect;
```
Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/createModel.js]: move the chai helper out to test/helpers/expect.js, then `var expect = require('../../../helpers/expect');`

**Comment by ritch** [comment]
@slnode test please

File: component/datamodel/graph.js (hunk 7 above)
Comment (deepakrkris) [inline component/datamodel/graph.js]: I want to export Graph as one of the exports, not the 'only' export

**Comment by superkhau** [comment]
@deepakrkris Please delete your branches after you merge.

**Pull request merged**